### ISSUE TITLE
How does Memories handle video time metadata?

### ISSUE THREAD

**Issue by Malakai-DB** [pulsejet/memories#1256]
Hello,

I opened this as a feature request as I am not sure if it is a bug and would need some clarifications first (couldn't find the information in the documentation).
I would like to understand how does Memories handle video time metadata.

I have the following situation : 
- I use multiple devices to capture pictures and videos, a bunch of Androids, a GoPro and a Nikon camera
- Everything is OK with timestamps when it comes to pictures, but for videos I see some weird behavior
- From what I understood, there isn't really a specification regarding the timestamps in a video metadata, but apparently the consensus is to have it in UTC format (correct me if I'm wrong)
- Android devices record the video timestamp in UTC, so Memories "adjusts" it according to the set timezone and the timeline looks good
- The GoPro and the Nikon camera record the video timestamp under my timezone
- For the GoPro, Memories "adjusts" it again and now all the GoPro videos are shifted in time in the timeline
- What is weird is that for the Nikon camera, Memories doesn't "adjust" the time and in the timeline I have it shown exactly at the time found in the video timestamp (which is correct)

I would like to understand this last behavior before trying to correct the GoPro timestamps in UTC.
In other words, I would like to know if the fact that Memories doesn't "adjust" the time for the Nikon videos is a bug, or there is an explanation for it.

Below are the exif data for one video taken with the GoPro and one with the Nikon. I couldn't identify anything that would make Memories think that the timestamp for the Nikon is in my timezone (as it doesn't "adjust" it).

<details>
  <summary>Nikon</summary>

Video taken at 2024:07:05 13:18:54 in my timezone and shown at the same time in Memories.
```
ExifTool Version Number         : 12.57
File Name                       : nikon.mov
Directory                       : .
File Size                       : 38 MB
File Modification Date/Time     : 2024:07:07 18:07:49+03:00
File Access Date/Time           : 2024:07:31 10:26:09+03:00
File Inode Change Date/Time     : 2024:07:08 02:13:03+03:00
File Permissions                : -rw-r--r--
File Type                       : MOV
File Type Extension             : mov
MIME Type                       : video/quicktime
Major Brand                     : Apple QuickTime (.MOV/QT)
Minor Version                   : 2007.9.0
Compatible Brands               : qt  , niko
Media Data Size                 : 37569256
Media Data Offset               : 32
Movie Header Version            : 0
Modify Date                     : 2024:07:05 13:18:54
Time Scale                      : 25000
Duration                        : 18.00 s
Preferred Rate                  : 1
Preferred Volume                : 100.00%
Preview Time                    : 0 s
Preview Duration                : 0 s
Poster Time                     : 0 s
Selection Time                  : 0 s
Selection Duration              : 0 s
Current Time                    : 0 s
Next Track ID                   : 3
Track Header Version            : 0
Track Create Date               : 2024:07:05 13:18:54
Track Modify Date               : 2024:07:05 13:18:54
Track ID                        : 1
Track Duration                  : 18.00 s
Track Layer                     : 0
Track Volume                    : 0.00%
Image Width                     : 1920
Image Height                    : 1080
Graphics Mode                   : srcCopy
Op Color                        : 0 0 0
Compressor ID                   : avc1
Source Image Width              : 1920
Source Image Height             : 1080
X Resolution                    : 72
Y Resolution                    : 72
Compressor Name                 : AVC Coding
Bit Depth                       : 24
Color Profiles                  : nclc
Color Primaries                 : BT.709
Transfer Characteristics        : BT.709
Matrix Coefficients             : BT.709
Video Frame Rate                : 25
Matrix Structure                : 1 0 0 0 1 0 0 0 1
Media Header Version            : 0
Media Create Date               : 2024:07:05 13:18:54
Media Modify Date               : 2024:07:05 13:18:54
Media Time Scale                : 48000
Media Duration                  : 18.00 s
Balance                         : 0
Handler Class                   : Data Handler
Handler Type                    : Alias Data
Audio Format                    : sowt
Maker Note Version              : Nikon 0.1.0.0
Make                            : NIKON
Model                           : COOLPIX L830
Software                        : COOLPIX L830 Ver.1.03
Create Date                     : 2024:07:05 13:18:54
Date/Time Original              : 2024:07:05 13:18:54
Frame Count                     : 450
Frame Rate                      : 0
Frame Width                     : 1920
Frame Height                    : 1080
Audio Channels                  : 2
Audio Bits Per Sample           : 16
Audio Sample Rate               : 48000
Thumbnail Image                 : (Binary data 10630 bytes, use -b option to extract)
Preview Image                   : (Binary data 51439 bytes, use -b option to extract)
Image Size                      : 1920x1080
Megapixels                      : 2.1
Avg Bitrate                     : 16.7 Mbps
Rotation                        : 0
```
</details>

<details>
  <summary>GoPro</summary>

Video taken at 2024:07:05 17:02:03 in my timezone but Memories "adjusts" it to my timezone (+3) so it shows it at 2024:07:05 20:02:03.
```
ExifTool Version Number         : 12.57
File Name                       : gopro.mp4
Directory                       : .
File Size                       : 186 MB
File Modification Date/Time     : 2024:07:07 17:53:37+03:00
File Access Date/Time           : 2024:07:30 20:04:46+03:00
File Inode Change Date/Time     : 2024:07:08 02:12:52+03:00
File Permissions                : -rw-r--r--
File Type                       : MP4
File Type Extension             : mp4
MIME Type                       : video/mp4
Major Brand                     : MP4 v1 [ISO 14496-1:ch13]
Minor Version                   : 2013.10.18
Compatible Brands               : mp41
Media Data Size                 : 185799036
Media Data Offset               : 28
Movie Header Version            : 0
Create Date                     : 2024:07:05 17:02:03
Modify Date                     : 2024:07:05 17:02:03
Time Scale                      : 60000
Duration                        : 0:00:49
Preferred Rate                  : 1
Preferred Volume                : 100.00%
Preview Time                    : 0 s
Preview Duration                : 0 s
Poster Time                     : 0 s
Selection Time                  : 0 s
Selection Duration              : 0 s
Current Time                    : 0 s
Next Track ID                   : 6
Firmware Version                : HD5.02.02.51.00
Lens Serial Number              : LAE7071814102047
Camera Serial Number Hash       : 04767991c6d194a4640d505237b014de
Track Header Version            : 0
Track Create Date               : 2024:07:05 17:02:03
Track Modify Date               : 2024:07:05 17:02:03
Track ID                        : 1
Track Duration                  : 0:00:49
Track Layer                     : 0
Track Volume                    : 0.00%
Image Width                     : 1920
Image Height                    : 1080
Graphics Mode                   : srcCopy
Op Color                        : 0 0 0
Compressor ID                   : avc1
Source Image Width              : 1920
Source Image Height             : 1080
X Resolution                    : 72
Y Resolution                    : 72
Compressor Name                 : GoPro AVC encoder
Bit Depth                       : 24
Color Profiles                  : nclx
Color Primaries                 : BT.709
Transfer Characteristics        : BT.709
Matrix Coefficients             : BT.709
Video Frame Rate                : 59.94
Time Code                       : 3
Balance                         : 0
Audio Format                    : mp4a
Audio Channels                  : 2
Audio Bits Per Sample           : 16
Audio Sample Rate               : 48000
Text Font                       : Unknown (21)
Text Face                       : Plain
Text Size                       : 10
Text Color                      : 0 0 0
Background Color                : 65535 65535 65535
Font Name                       : Helvetica
Other Format                    : tmcd
Playback Frame Rate             : 59.94005994
Warning                         : [minor] The ExtractEmbedded option may find more tags in the media data
Matrix Structure                : 1 0 0 0 1 0 0 0 1
Media Header Version            : 0
Media Create Date               : 2024:07:05 17:02:03
Media Modify Date               : 2024:07:05 17:02:03
Media Time Scale                : 60000
Media Duration                  : 0:00:49
Handler Class                   : Media Handler
Handler Type                    : NRT Metadata
Handler Description             : GoPro SOS
Gen Media Version               : 0
Gen Flags                       : 0 0 0
Gen Graphics Mode               : srcCopy
Gen Op Color                    : 0 0 0
Gen Balance                     : 0
Meta Format                     : fdsc
Image Size                      : 1920x1080
Megapixels                      : 2.1
Avg Bitrate                     : 30.1 Mbps
Rotation                        : 0
```
</details>

**Comment by pulsejet** [comment]
tl;dr it's a bad mess 

https://github.com/pulsejet/memories/issues/815#issuecomment-1741793109

**Comment by Malakai-DB** [comment]
That is what I also understood after making some research.
What bothers me in this case is the difference of behavior of Memories between the Nikon videos and the GoPro/Android videos.

I am not a developer, nor I know php, but looking at this part of the code (https://github.com/pulsejet/memories/blob/002c2111a61516da2c631f7e6f8a28b5671161aa/lib/Exif.php#L157C1-L179C56), it seems it tries to get the timezone from "OffsetTimeOriginal", "OffsetTime", "LocationTZID", which none of the videos from the Nikon or the GoPro have (as can be seen on the first post).
And thus it should force it to UTC timezone according to the line 179 (which it does for Android/GoPro videos BUT NOT for Nikon videos).

So why does Memories not suppose (like for the Android and GoPro videos) that the timezone the Nikon videos were taken is UTC?

**Comment by pulsejet** [comment]
Perhaps some fields are missing here? Are you using these args?

https://github.com/pulsejet/memories/blob/002c2111a61516da2c631f7e6f8a28b5671161aa/lib/Exif.php#L15

**Comment by Malakai-DB** [comment]
Thank you for your previous comment. That allowed me to investigate further.
Like we can see from the output below, I tried to reproduce what Memories might do to extract the time from the video files.
It seems that exiftool, when appending the parameters "-api QuickTimeUTC=1" and "-G1", outputs 2 "CreateDate" for the Nikon videos, but only one for the GoPro.
If the group name parameter (-G1) is not used, then there is no real change for the GoPro, but for the Nikon, we get the "Nikon:CreateDate" and not the "QuickTime:CreateDate". I suppose exiftool considers "QuickTime:CreateDate" to be in UTC but doesn't really bother with the "Nikon:CreateDate".
As "Nikon:CreateDate" is not "adjusted" by exiftool and is outputted as is, Memories not seeing the offset (+03:00), suppose the video was made in its own timezone, therefore it is correctly shown in the timeline.
I also just remembered that the Nikon camera, when first started after putting the batteries in, asks to set the timezone and the time (even if I am not sure that it writes the timezone anywhere, as I was not able to find this information in the metadata), whereas the GoPro doesn't have the possibility to add a timezone (only to set the time).
Nevertheless, this difference in behavior between the Nikon and the GoPro videos are inherent to exiftool and not Memories (as the latter relies on the first to get the metadata).
So why doesn't exiftool "adjust" also the "Nikon:CreateDate" besides the "QuickTime:CreateDate", and why does it prefer to display the "Nikon:CreateDate" instead of the "QuickTime:CreateDate" when the group name parameter (-G1) is missing, is a total mistery (maybe it just displays the last "CreateDate" information in the metadata and for the Nikon it seems to always be "Nikon:CreateDate").
In my case, as long as exiftool will by default prefer to output the "Nikon:CreateDate" (without adjusting it) instead of the "QuickTime:CreateDate", I think I will be safe for the Nikon videos already present and the future ones, as they will be correctly displayed in the Memories timeline. As for the GoPro ones, it seems I will have to shift the time before adding any new video to Memories.


```
$ exiftool -AllDates -api QuickTimeUTC=1 -api LargeFileSupport=1 -n -json -G1 nikon.mov 
[{
  "SourceFile": "nikon.mov",
  "Nikon:DateTimeOriginal": "2024:07:05 13:18:54",
  "QuickTime:CreateDate": "2024:07:05 16:18:54+03:00",
  "Nikon:CreateDate": "2024:07:05 13:18:54",
  "QuickTime:ModifyDate": "2024:07:05 16:18:54+03:00"
}]

$ exiftool -AllDates -api QuickTimeUTC=1 -api LargeFileSupport=1 -n -json nikon.mov 
[{
  "SourceFile": "nikon.mov",
  "DateTimeOriginal": "2024:07:05 13:18:54",
  "CreateDate": "2024:07:05 13:18:54",
  "ModifyDate": "2024:07:05 16:18:54+03:00"
}]


$ exiftool -AllDates -api QuickTimeUTC=1 -api LargeFileSupport=1 -n -json -G1 gopro.mp4
[{
  "SourceFile": "gopro.mp4",
  "QuickTime:CreateDate": "2024:07:05 20:02:03+03:00",
  "QuickTime:ModifyDate": "2024:07:05 20:02:03+03:00"
}]

$ exiftool -AllDates -api QuickTimeUTC=1 -api LargeFileSupport=1 -n -json gopro.mp4
[{
  "SourceFile": "gopro.mp4",
  "CreateDate": "2024:07:05 20:02:03+03:00",
  "ModifyDate": "2024:07:05 20:02:03+03:00"
}]
```


@pulsejet : Thank you for your help!
